pkg/streamer/file.go

```diff
@@ -186,141 +187,90 @@ func fileSizeUpdated(path string, latestSize int64) (int, error) {
+		log.L().Error("size of relay log file has been changed", zap.String("file", path),
+			zap.Int64("old size", latestSize), zap.Int64("size", curSize))
 		return -1, nil
 	}
 }
 
-// relaySubDirUpdated checks whether the relay sub directory updated
-// including file changed, created, removed, etc.
-func relaySubDirUpdated(ctx context.Context, watcherInterval time.Duration, dir string,
+// relayLogUpdatedOrNewCreated checks whether current relay log file is updated or new relay log is created.
+// we check the size of the file first, if the size is the same as the latest file, we assume there is no new write
+// so we need to check relay meta file to see if the new relay log is created.
+// this func will be blocked until current filesize changed or meta file updated or context cancelled.
+func relayLogUpdatedOrNewCreated(ctx context.Context, watcherInterval time.Duration, dir string,
 	latestFilePath, latestFile string, latestFileSize int64, updatePathCh chan string, errCh chan error) {
-	var err error
-	defer func() {
-		if err != nil {
-			errCh <- err
-		}
-	}()
-
-	// create polling watcher
-	watcher2 := watcher.NewWatcher()
-
-	// Add before Start
-	// no need to Remove, it will be closed and release when return
-	err = watcher2.Add(dir)
-	if err != nil {
-		err = terror.ErrAddWatchForRelayLogDir.Delegate(err, dir)
-		return
-	}
-
-	err = watcher2.Start(watcherInterval)
-	if err != nil {
-		err = terror.ErrWatcherStart.Delegate(err, dir)
-		return
-	}
-	defer watcher2.Close()
-
-	type watchResult struct {
-		updatePath string
-		err        error
-	}
-
-	result := make(chan watchResult, 1) // buffered chan to ensure not block the sender even return in the halfway
-	newCtx, cancel := context.WithCancel(ctx)
-	defer cancel()
-	go func() {
-		for {
-			select {
-			case <-newCtx.Done():
-				result <- watchResult{
-					updatePath: "",
-					err:        newCtx.Err(),
-				}
+	ticker := time.NewTicker(watcherInterval)
+	defer ticker.Stop()
+	for {
+		select {
+		case <-ctx.Done():
+			errCh <- terror.Annotate(ctx.Err(), "context meet error")
+			return
+		case <-ticker.C:
+			// check the latest relay log file whether updated when adding watching and collecting newer
+			cmp, err := fileSizeUpdated(latestFilePath, latestFileSize)
+			if err != nil {
+				errCh <- terror.Annotatef(err, "latestFilePath=%s latestFileSize=%d", latestFilePath, latestFileSize)
 				return
-			case err2, ok := <-watcher2.Errors:
-				if !ok {
-					result <- watchResult{
-						updatePath: "",
-						err:        terror.ErrWatcherChanClosed.Generate("errors", dir),
-					}
-				} else {
-					result <- watchResult{
-						updatePath: "",
-						err:        terror.ErrWatcherChanRecvError.Delegate(err2, dir),
-					}
-				}
+			}
+			failpoint.Inject("CMPAlwaysReturn0", func(_ failpoint.Value) {
+				cmp = 0
+			})
+			switch {
+			case cmp < 0:
```

Comment:
What if we clean some relay logs?

Reply:
`fileSizeUpdated` will return a error and task will pause

Reply:
What about `relayPurger`? Will this cause the task to pause?

Reply:
relayPurger does not affect the task 

because it does not clear the protected relaylog， see more in `setActiveRelayLog`